portal/flatpak-portal.c

```diff
@@ -1024,6 +1026,124 @@ handle_spawn_signal (PortalFlatpak         *object,
+        {
+          g_auto(GStrv) fields = NULL;
+          guint nfields = 0;
+          char *endptr = NULL;
+
+          fields = g_strsplit (line, "\t", -1);
+          nfields = g_strv_length (fields);
+          if (nfields < 3)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                           "NSpid line has too few fields: %s", line);
+              return 0;
+            }
+
+          /* The second to last PID namespace is the one that spawned this process */
+          relative_pid = strtol (fields[nfields - 2], &endptr, 10);
+          if (*endptr)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                           "Invalid parent-relative PID in NSpid line: %s", line);
+              return 0;
+            }
+
+          return relative_pid;
+        }
+    }
+
+  if (*error == NULL)
+    /* EOF was reached while reading the file */
+    g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "NSpid not found");
+
+  return 0;
+}
+
+static gboolean
+handle_spawn_relative_pid (PortalFlatpak         *object,
+                           GDBusMethodInvocation *invocation,
+                           guint                  arg_pid)
+{
+  PidData *pid_data = NULL;
+  g_autoptr(GPtrArray) instances = NULL;
+
+  g_debug ("spawn_relative_pid(%d)", arg_pid);
+
+  pid_data = g_hash_table_lookup (client_pid_data_hash, GUINT_TO_POINTER (arg_pid));
+  if (pid_data == NULL || !pid_data->expose_pids ||
+      strcmp (pid_data->client, g_dbus_method_invocation_get_sender (invocation)) != 0)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR,
+                                             G_DBUS_ERROR_UNIX_PROCESS_ID_UNKNOWN,
+                                             "No such pid");
+      return TRUE;
+    }
+
+  instances = flatpak_instance_get_all ();
+  for (int i = 0; i < instances->len; i++)
+    {
+      FlatpakInstance *instance = g_ptr_array_index (instances, i);
+      if (flatpak_instance_get_pid (instance) == arg_pid)
+        {
+          g_autoptr(GError) error = NULL;
+          int child_pid, relative_pid;
+
+          child_pid = flatpak_instance_get_child_pid (instance);
+          /* If child_pid == 0, most likely the child process has not been
+             started yet, resulting in an empty or invalid bwrapinfo.json.
+             Just pretend the instance itself wasn't found, so the caller
+             can handle them in the same way. */
+          if (child_pid == 0)
+            break;
+
+          relative_pid = get_child_pid_relative_to_parent_sandbox (child_pid, &error);
+          if (relative_pid == 0)
+            {
+              g_dbus_method_invocation_return_gerror (invocation, error);
+              return TRUE;
+            }
+
+          portal_flatpak_complete_spawn_relative_pid (portal, invocation, relative_pid);
+
+          return TRUE;
+        }
+    }
```

Comment:
Iterating over all instances like this seems quite expensive and fragile. I think a better option would be to add an option to flatpak run to write the instance it to a file descriptor, and then record this in the PidData.

Reply:
Hmm my original thought was actually to add an argument that corresponds to bwrap's `--json-status-fd`, then just emit a `SpawnSuccess` signal with the PID relative to the spawner's namespace, which would be a bit prettier overall. Would that work? I just wasn't certain about the upstream feel for adding fd-related arguments to `flatpak run` itself...

Reply:
FWIW in terms of the use case (Chromium) this would work just as well.